test/unit-test/sigv4_utest.c

```diff
@@ -214,3 +523,468 @@ void test_SigV4_AwsIotDateToIso8601_Formatting_Error()
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+}
+
+/* Test the API for handling corner cases of sorting the Query Parameters (when generating Canonical Query) */
+void test_SigV4_GenerateHTTPAuthorization_Sorting_Query_Params_Corner_Cases()
+{
+    SigV4Status_t returnStatus;
+
+    /* Test when the query string contains query parameters with exactly matching names. */
+    params.pHttpParameters->pQuery = QUERY_MATCHING_PARAMS;
+    params.pHttpParameters->queryLen = strlen( QUERY_MATCHING_PARAMS );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* Test when the query string contains query parameters which do not match in name length but match in
+     * the name data for the common length between 2 parameters. */
+    params.pHttpParameters->pQuery = QUERY_WITH_MATCHING_PARAM_PREFIX;
+    params.pHttpParameters->queryLen = strlen( QUERY_WITH_MATCHING_PARAM_PREFIX );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* Test when the query string contains query parameters with exactly matching parameter names as well
+     * as matching values for those parameters. The query values of matching parameters differ in length though. */
+    params.pHttpParameters->pQuery = QUERY_MATCHING_PARAMS_AND_MATCHING_VALUES_PREFIX_1;
+    params.pHttpParameters->queryLen = strlen( QUERY_MATCHING_PARAMS_AND_MATCHING_VALUES_PREFIX_1 );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+    params.pHttpParameters->pQuery = QUERY_MATCHING_PARAMS_AND_MATCHING_VALUES_PREFIX_2;
+    params.pHttpParameters->queryLen = strlen( QUERY_MATCHING_PARAMS_AND_MATCHING_VALUES_PREFIX_2 );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    /* Test when the query string contains query parameters with exactly matching parameter names as well
+     * as matching values for those parameters. The query values of matching parameters differ in length though. */
+    params.pHttpParameters->pQuery = QUERY_MATCHING_PARAMS_AND_VALUES;
+    params.pHttpParameters->queryLen = strlen( QUERY_MATCHING_PARAMS_AND_VALUES );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+}
+
+/* Test that the library can handle HTTP headers data containing trimmable spaces. */
+void test_SigV4_GenerateAuthorization_Headers_With_Trimmable_Spaces()
```

Comment:
Is there any way to verify the headers were trimmed properly?

Reply:
That's a good question. The only way in the current unit testing structure of only testing the public APIs is to test the signature generated. I am thinking of creating a separate PR for verifying the generated signatures for tests that expect successful generation of signature.